sql/flowinfra: panic in rangefeed.(*CatchUpIterator)

sql/flowinfra.TestEvalCtxTxnOnRemoteNodes [failed](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/12593727?buildTab=log) with [artifacts](https://teamcity.cockroachdb.com/buildConfiguration/Cockroach_Nightlies_StressBazel/12593727?buildTab=artifacts#/) on release-23.2 @ [5c59f412f63925f930a58392d0a95980b37e0655](https://github.com/cockroachdb/cockroach/commits/5c59f412f63925f930a58392d0a95980b37e0655):

Fatal error:

```
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x19a pc=0x1cb6389]

```
Stack: 

```
goroutine 4989621 [running]:
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).recover(0xc00e3327f0?, {0x649f108, 0xc001173110})
	github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:230 +0x6a
panic({0x49d0f80, 0x8fcedc0})
	GOROOT/src/runtime/panic.go:884 +0x213
github.com/cockroachdb/cockroach/pkg/storage.(*pebbleIterator).Close(0xc00e332b60?)
	github.com/cockroachdb/cockroach/pkg/storage/pebble_iterator.go:338 +0x29
github.com/cockroachdb/cockroach/pkg/storage.(*intentInterleavingIter).Close(0xc01318fb00)
	github.com/cockroachdb/cockroach/pkg/storage/intent_interleaving_iter.go:1068 +0x3d
github.com/cockroachdb/cockroach/pkg/storage.(*MVCCIncrementalIterator).Close(0xc0068d96c0)
	github.com/cockroachdb/cockroach/pkg/storage/mvcc_incremental_iterator.go:278 +0x29
github.com/cockroachdb/cockroach/pkg/kv/kvserver/rangefeed.(*CatchUpIterator).Close(0xc009976540)
	github.com/cockroachdb/cockroach/pkg/kv/kvserver/rangefeed/catchup_scan.go:114 +0x2d
github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Replica).registerWithRangefeedRaftMuLocked.func1()
	github.com/cockroachdb/cockroach/pkg/kv/kvserver/pkg/kv/kvserver/replica_rangefeed.go:395 +0x2a
github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Replica).registerWithRangefeedRaftMuLocked(0xc012936c80, {0x649f108, 0xc01093a420}, {{0xc00daf16d0, 0x3, 0x8}, {0xc00daf16d8, 0x3, 0x8}}, {0x179611e806e4d03b, ...}, ...)
	github.com/cockroachdb/cockroach/pkg/kv/kvserver/pkg/kv/kvserver/replica_rangefeed.go:485 +0xbaf
github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Replica).RangeFeed(0xc012936c80, 0xc0197cb1e0, {0x6475a48, 0xc01093a2d0}, 0x3?)
	github.com/cockroachdb/cockroach/pkg/kv/kvserver/pkg/kv/kvserver/replica_rangefeed.go:288 +0x6b0
github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Store).RangeFeed(0xc004097500, 0xc0197cb1e0, {0x6475a48, 0xc01093a2d0})
	github.com/cockroachdb/cockroach/pkg/kv/kvserver/pkg/kv/kvserver/store.go:3005 +0x105
github.com/cockroachdb/cockroach/pkg/kv/kvserver.(*Stores).RangeFeed(0xc000e12c80?, 0xc0197cb1e0, {0x6475a48, 0xc01093a2d0})
	github.com/cockroachdb/cockroach/pkg/kv/kvserver/pkg/kv/kvserver/stores.go:227 +0xf6
github.com/cockroachdb/cockroach/pkg/server.(*Node).MuxRangeFeed(0xc0093cb000, {0x64ddb68?, 0xc01c227f80})
	github.com/cockroachdb/cockroach/pkg/server/node.go:1846 +0x5a6
github.com/cockroachdb/cockroach/pkg/kv/kvpb._Internal_MuxRangeFeed_Handler({0x5066540?, 0xc0093cb000}, {0x64d1e70?, 0xc000e129a0})
	github.com/cockroachdb/cockroach/pkg/kv/kvpb/bazel-out/k8-fastbuild/bin/pkg/kv/kvpb/kvpb_go_proto_/github.com/cockroachdb/cockroach/pkg/kv/kvpb/api.pb.go:10457 +0x9f
github.com/cockroachdb/cockroach/pkg/util/tracing/grpcinterceptor.StreamServerInterceptor.func1({0x5066540, 0xc0093cb000}, {0x64ce2d0?, 0xc0011739e0?}, 0xc0164e8ff0, 0x5433c90)
	github.com/cockroachdb/cockroach/pkg/util/tracing/grpcinterceptor/grpc_interceptor.go:165 +0x6c4
google.golang.org/grpc.getChainStreamHandler.func1({0x5066540, 0xc0093cb000}, {0x64ce2d0, 0xc0011739e0})
	google.golang.org/grpc/external/org_golang_google_grpc/server.go:1478 +0xb9
github.com/cockroachdb/cockroach/pkg/rpc.NewServerEx.func4({0x5066540, 0xc0093cb000}, {0x64ce2d0, 0xc0011739e0}, 0xc0164e8ff0?, 0xc00df13480)
	github.com/cockroachdb/cockroach/pkg/rpc/pkg/rpc/context.go:178 +0x74
google.golang.org/grpc.getChainStreamHandler.func1({0x5066540, 0xc0093cb000}, {0x64ce2d0, 0xc0011739e0})
	google.golang.org/grpc/external/org_golang_google_grpc/server.go:1478 +0xb9
github.com/cockroachdb/cockroach/pkg/rpc.kvAuth.streamInterceptor({0xc00c990000?, {{0x4398c7?}, {0x64d6b70?, 0xc01a006450?}}}, {0x5066540, 0xc0093cb000}, {0x64d24e8, 0xc004951c00}, 0xc0164e8ff0, 0xc00df13340)
	github.com/cockroachdb/cockroach/pkg/rpc/pkg/rpc/auth.go:167 +0x3b1
google.golang.org/grpc.getChainStreamHandler.func1({0x5066540, 0xc0093cb000}, {0x64d24e8, 0xc004951c00})
	google.golang.org/grpc/external/org_golang_google_grpc/server.go:1478 +0xb9
github.com/cockroachdb/cockroach/pkg/rpc.NewServerEx.func2.1({0xc00c7c2480?, 0x441970?})
	github.com/cockroachdb/cockroach/pkg/rpc/pkg/rpc/context.go:145 +0x2d
github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).RunTaskWithErr(0xc00c7c2480, {0x649f108, 0xc001173110}, {0xc00043e000?, 0xc00df12cc0?}, 0xc00bc9daa0)
	github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:336 +0xd1
github.com/cockroachdb/cockroach/pkg/rpc.NewServerEx.func2({0x5066540, 0xc0093cb000}, {0x64d24e8?, 0xc004951c00?}, 0xc0164e8ff0, 0xc00df12cc0)
	github.com/cockroachdb/cockroach/pkg/rpc/pkg/rpc/context.go:144 +0xef
google.golang.org/grpc.chainStreamInterceptors.func1({0x5066540, 0xc0093cb000}, {0x64d24e8, 0xc004951c00}, 0x47a0300?, 0xc01c298000?)
	google.golang.org/grpc/external/org_golang_google_grpc/server.go:1469 +0x8f
google.golang.org/grpc.(*Server).processStreamingRPC(0xc00fcde1e0, {0x64e98a0, 0xc019189520}, 0xc0160d27e0, 0xc005a48780, 0x9011f00, 0x0)
	google.golang.org/grpc/external/org_golang_google_grpc/server.go:1627 +0x1206
google.golang.org/grpc.(*Server).handleStream(0xc00fcde1e0, {0x64e98a0, 0xc019189520}, 0xc0160d27e0, 0x0)
	google.golang.org/grpc/external/org_golang_google_grpc/server.go:1708 +0x9f0
google.golang.org/grpc.(*Server).serveStreams.func1.2()
	google.golang.org/grpc/external/org_golang_google_grpc/server.go:965 +0x98
created by google.golang.org/grpc.(*Server).serveStreams.func1
	google.golang.org/grpc/external/org_golang_google_grpc/server.go:963 +0x28a
```
<details><summary>Log preceding fatal error</summary>
<p>

```
=== RUN   TestEvalCtxTxnOnRemoteNodes
    test_log_scope.go:170: test logs captured to: /artifacts/tmp/_tmp/fd023fd6c8beefd8db32fbb739554fec/logTestEvalCtxTxnOnRemoteNodes2567333182
    test_log_scope.go:81: use -show-logs to present logs inline
    test_server_shim.go:165: automatically injected a shared process virtual cluster under test; see comment at top of test_server_shim.go for details.
=== RUN   TestEvalCtxTxnOnRemoteNodes/vectorize=false
=== RUN   TestEvalCtxTxnOnRemoteNodes/vectorize=true
```
</p>
</details>
<p>Parameters: <code>TAGS=bazel,gss,deadlock</code>
, <code>stress=true</code>
</p>
<details><summary>Help</summary>
<p>

See also: [How To Investigate a Go Test Failure \(internal\)](https://cockroachlabs.atlassian.net/l/c/HgfXfJgM)
</p>
</details>
/cc @cockroachdb/sql-queries
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestEvalCtxTxnOnRemoteNodes.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>


Jira issue: CRDB-33388

--- a/pkg/kv/kvserver/rangefeed/scheduled_processor.go
+++ b/pkg/kv/kvserver/rangefeed/scheduled_processor.go
@@ -18,6 +18,7 @@ import (
 	"github.com/cockroachdb/cockroach/pkg/kv/kvpb"
 	"github.com/cockroachdb/cockroach/pkg/roachpb"
 	"github.com/cockroachdb/cockroach/pkg/storage/enginepb"
+	"github.com/cockroachdb/cockroach/pkg/util/buildutil"
 	"github.com/cockroachdb/cockroach/pkg/util/future"
 	"github.com/cockroachdb/cockroach/pkg/util/hlc"
 	"github.com/cockroachdb/cockroach/pkg/util/log"
@@ -574,19 +575,37 @@ func (p *ScheduledProcessor) Filter() *Filter {
 // is guaranteed that only single request is modifying processor at any given
 // time. It is advisable to use provided processor reference for operations
 // rather than using one within closure itself.
-// If request can't be queued or processor stoppedC is closed then default
-// value is returned.
+//
+// If the processor is stopped concurrently with the request queueing, it may or
+// may not be processed. If the request is ever processed, its return value is
+// guaranteed to be returned here. Otherwise, the zero value is returned and the
+// request is never processed.
 func runRequest[T interface{}](
 	p *ScheduledProcessor, f func(ctx context.Context, p *ScheduledProcessor) T,
 ) (r T) {
 	result := make(chan T, 1)
 	p.enqueueRequest(func(ctx context.Context) {
 		result <- f(ctx, p)
+		// Assert that we never process requests after stoppedC is closed. This is
+		// necessary to coordinate catchup iter ownership and avoid double-closing.
+		if buildutil.CrdbTestBuild {
+			select {
+			case <-p.stoppedC:
+				log.Fatalf(ctx, "processing request on stopped processor")
+			default:
+			}
+		}
 	})
 	select {
 	case r = <-result:
 		return r
 	case <-p.stoppedC:
+		// If the request was processed concurrently with a stop, there's a 50%
+		// chance we didn't take the result branch. Check again.
+		select {
+		case r = <-result:
+		default:
+		}
 		return r
 	}
 }

